lib/cfg.js

```diff
@@ -0,0 +1,368 @@
+            fnRange.start = first;
+        }
+        ++first;
+    }
+
+    fnRange.end = last;
+    result.push(fnRange);//possiblitiy of bug?! because didn't clone(JSON.parse(JSON.stringify(...))
+    return result;
+};
+
+ControlFlowGraph.prototype.splitToBasicBlocks = function (asmArr,  range, isEnd, isJmp) {
+    var first = range.start;
+    var last = range.end;
+    if(first == last) return [];
+    ++first;
+
+    var rangeBb = {nameId:"start" , start:first , end:null, actionPos:[]};
+    var result = [];
+
+    var resetRangeWith = function(rangeBb,  nameId, start) {
+            rangeBb.nameId = nameId;//.L1:
+            rangeBb.start = start;//after .L1:
+            rangeBb.actionPos = [];
+    };
+
+    while(first != last) {
+        var inst = asmArr[first].text;
+        if(isEnd(inst)) {
+            rangeBb.end = first;
+            result.push(JSON.parse(JSON.stringify(rangeBb)));
+            ++first;//risk of a bug.
+            //inst is expected to be .L*: where * in 1,2,...
+            resetRangeWith(rangeBb, inst, first);
+        }
+        else if(isJmp(inst)) {
+                rangeBb.actionPos.push(first);
+        }
+        ++first;
+    }
+
+    rangeBb.end = last;
+    result.push(JSON.parse(JSON.stringify(rangeBb)));
+    return result;
+};
+
+ControlFlowGraph.prototype.splitToCanonicalBasicBlock = function (basicBlock) {
+    var actionPos = basicBlock.actionPos;
+    var actPosSz = actionPos.length;
+    if (actionPos[actPosSz-1] + 1 == basicBlock.end){
+        --actPosSz;
+    }
+
+    if(actPosSz === 0)
+        return [{
+                    nameId: basicBlock.nameId,
+                    start: basicBlock.start,
+                    end: basicBlock.end
+               }];
+    else if(actPosSz == 1)
+            return [
+                {nameId: basicBlock.nameId,start: basicBlock.start,end:actionPos[0]+1},
+                {nameId: basicBlock.nameId+"@"+ (actionPos[0]+1),start: actionPos[0]+1,end: basicBlock.end}
+               ];
+    else {
+        var first = 0;
+        var last = actPosSz;
+        var block_name = basicBlock.nameId;
+        var tmp = {nameId:block_name, start:basicBlock.start, end:actionPos[first]+1 };
+        var result = [];
+        result.push(JSON.parse(JSON.stringify(tmp)));
+        while(first != last-1) {
+            tmp.nameId = block_name + "@" + (actionPos[first]+1);
+            tmp.start = actionPos[first]+1;
+            ++first;
+            tmp.end = actionPos[first]+1;
+            result.push(JSON.parse(JSON.stringify(tmp)));
+        }
+        
+        tmp = {nameId:block_name + "@" + (actionPos[first] + 1), start:actionPos[first] + 1, end:basicBlock.end };
+        result.push(JSON.parse(JSON.stringify(tmp)));
+        
+        return result;
+
+    }
+
+};
+
+ControlFlowGraph.prototype.concatInstructions = function (asmArr, first, last) {
+    
+    var makeShorter = function(instruction){
+        if(instruction.length > 50){
```

Comment:
`substr` never returns a longer string than it started with so you can replace this with

```
function makeShorter(instruction) { return instruction.substr(0, 50); }
```

Reply:
done